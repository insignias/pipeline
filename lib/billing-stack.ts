import { Stack, StackProps } from "aws-cdk-lib";
import { Construct } from "constructs";
import { Budget } from "./Construct/budget";

interface BillingStackProps extends StackProps{
    amount: number,
    emailAddress: string
}

export class BillingStack extends Stack{
    constructor(scope: Construct, id: string, props: BillingStackProps){
        super(scope, id, props);

        new Budget(this, 'Budget', {
            amount: props.amount,
            emailAddress: props.emailAddress
        })
    }
}