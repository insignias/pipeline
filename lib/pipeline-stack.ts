import { SecretValue, Stack, StackProps } from 'aws-cdk-lib';
import { BuildSpec, LinuxBuildImage, PipelineProject } from 'aws-cdk-lib/lib/aws-codebuild';
import { Artifact, Pipeline } from 'aws-cdk-lib/lib/aws-codepipeline';
import { CloudFormationCreateUpdateStackAction, CodeBuildAction, GitHubSourceAction } from 'aws-cdk-lib/lib/aws-codepipeline-actions';
import { Construct } from 'constructs';

export class PipelineStack extends Stack {
  private pipeline: Pipeline;
  private cdkBuildOutput: Artifact;

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    this.pipeline = new Pipeline(this, "Pipeline", {
      pipelineName: "Pipeline",
      crossAccountKeys: true,
      restartExecutionOnUpdate: true,
    });

    const cdkSourceOutput = new Artifact("CDKSourceOutput");

    this.pipeline.addStage({
      stageName: "Source",
      actions: [
        new GitHubSourceAction({
          owner: "insignias",
          repo: "pipeline",
          branch: "main",
          actionName: "Pipeline_Source",
          oauthToken: SecretValue.secretsManager("github-token"),
          output: cdkSourceOutput,
        })
      ]
    });

    this.cdkBuildOutput = new Artifact("CdkBuildOutput");

    this.pipeline.addStage({
      stageName: "Build",
      actions: [
        new CodeBuildAction({
          actionName: "CDK_Build",
          input: cdkSourceOutput,
          outputs: [this.cdkBuildOutput],
          project: new PipelineProject(this, "CdkBuildProject", {
            environment: {
              buildImage: LinuxBuildImage.STANDARD_5_0,
            },
            buildSpec: BuildSpec.fromSourceFilename(
              "build-specs/cdk-build-spec.yml"
            ),
          }),
        })
      ]
    });

    this.pipeline.addStage({
      stageName: "Pipeline_Update",
      actions: [
        new CloudFormationCreateUpdateStackAction({
          actionName: "Pipeline_Update",
          stackName: "PipelineStack",
          templatePath: this.cdkBuildOutput.atPath(
            "PipelineStack.template.json"
          ),
          adminPermissions: true,
        }),
      ],
    });
  }
}
