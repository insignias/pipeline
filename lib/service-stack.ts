import { CfnParameter, Stack, StackProps } from "aws-cdk-lib";
import { CfnApi, CfnIntegration, HttpApi } from "aws-cdk-lib/lib/aws-apigatewayv2";
import { LambdaProxyIntegration } from "aws-cdk-lib/lib/aws-apigatewayv2-integrations";
import { CfnParametersCode, CfnPermission, Code, Function, Runtime } from "aws-cdk-lib/lib/aws-lambda";
import { Construct } from "constructs";

export class ServiceStack extends Stack {
    public readonly serviceCode: CfnParametersCode;

    constructor(scope: Construct, id: string, props?: StackProps) {
        super(scope, id, props);

        this.serviceCode = Code.fromCfnParameters();

        const lambda = new Function(this, 'ServiceLambda', {
            runtime: Runtime.NODEJS_14_X,
            handler: 'src/lambda.handler',
            code: this.serviceCode,
            functionName: 'ServiceLambda'
        })

        // const api = new CfnApi(this, 'ServiceApi', {
        //     protocolType: 'HTTP',
        //     name: 'ServiceApi',
        //     target: lambda.functionArn
        // });

        // const permission = new CfnPermission(this, 'lambdaPermission', {
        //     functionName: lambda.functionArn,
        //     action: 'lambda:InvokeFunction',
        //     principal: 'apigateway.amazonaws.com',
        //     sourceArn: api.
        // })

        
        new HttpApi(this, 'ServiceApi', {
            defaultIntegration: new LambdaProxyIntegration({
                handler: lambda
            }),
            apiName: 'MyService'
        })
    }
}
