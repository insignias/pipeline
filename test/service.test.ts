import { App} from 'aws-cdk-lib';
import { ServiceStack} from '../lib/service-stack';
import { expect as expectCDK, SynthUtils} from '@aws-cdk/assert';

test('ServiceStackTest', () => {
    const app = new App();
    // WHEN
    const stack = new ServiceStack(app, 'ServiceStackTest', {stageName: 'dummy'});
    // THEN
    expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});