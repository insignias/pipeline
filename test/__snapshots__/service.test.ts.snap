// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ServiceStackTest 1`] = `
Object {
  "Outputs": Object {
    "ServiceEndpoint": Object {
      "Description": "ServiceEndpoint",
      "Export": Object {
        "Name": "ServiceEndpoint",
      },
      "Value": Object {
        "Fn::GetAtt": Array [
          "ServiceApi4633ACD6",
          "ApiEndpoint",
        ],
      },
    },
  },
  "Parameters": Object {
    "ServiceLambdaLambdaSourceBucketNameParameterA6290FE3": Object {
      "Type": "String",
    },
    "ServiceLambdaLambdaSourceObjectKeyParameter86BEE2BA": Object {
      "Type": "String",
    },
  },
  "Resources": Object {
    "ServiceApi4633ACD6": Object {
      "Properties": Object {
        "Name": "MyServicedummy",
        "ProtocolType": "HTTP",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
    "ServiceApiDefaultRoute1D980EEE": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "ServiceApi4633ACD6",
        },
        "AuthorizationType": "NONE",
        "RouteKey": "$default",
        "Target": Object {
          "Fn::Join": Array [
            "",
            Array [
              "integrations/",
              Object {
                "Ref": "ServiceApiDefaultRouteHttpIntegration017fdbd76b07f361329d8f77e9cebcdb4F69854C",
              },
            ],
          ],
        },
      },
      "Type": "AWS::ApiGatewayV2::Route",
    },
    "ServiceApiDefaultRouteHttpIntegration017fdbd76b07f361329d8f77e9cebcdb4F69854C": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "ServiceApi4633ACD6",
        },
        "IntegrationType": "AWS_PROXY",
        "IntegrationUri": Object {
          "Fn::GetAtt": Array [
            "ServiceLambdaA7DE4D2C",
            "Arn",
          ],
        },
        "PayloadFormatVersion": "2.0",
      },
      "Type": "AWS::ApiGatewayV2::Integration",
    },
    "ServiceApiDefaultRouteServiceStackTestServiceApiDefaultRouteF560CF2APermission2ACE02F1": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "ServiceLambdaA7DE4D2C",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": Object {
          "Fn::Join": Array [
            "",
            Array [
              "arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              Object {
                "Ref": "AWS::Region",
              },
              ":",
              Object {
                "Ref": "AWS::AccountId",
              },
              ":",
              Object {
                "Ref": "ServiceApi4633ACD6",
              },
              "/*/*",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ServiceApiDefaultStage0A39469D": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "ServiceApi4633ACD6",
        },
        "AutoDeploy": true,
        "StageName": "$default",
      },
      "Type": "AWS::ApiGatewayV2::Stage",
    },
    "ServiceLambdaA7DE4D2C": Object {
      "DependsOn": Array [
        "ServiceLambdaServiceRoleF5A50FBC",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "ServiceLambdaLambdaSourceBucketNameParameterA6290FE3",
          },
          "S3Key": Object {
            "Ref": "ServiceLambdaLambdaSourceObjectKeyParameter86BEE2BA",
          },
        },
        "FunctionName": "ServiceLambdadummy",
        "Handler": "src/lambda.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "ServiceLambdaServiceRoleF5A50FBC",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "ServiceLambdaServiceRoleF5A50FBC": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
